vms/evm/metrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package prometheus
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/ava-labs/libevm/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Gatherer implements [prometheus.Gatherer] interface by
+// gathering all metrics from the given Prometheus registry.
+type Gatherer struct {
+	registry Registry
+}
+
+var _ prometheus.Gatherer = (*Gatherer)(nil)
+
+// NewGatherer returns a [Gatherer] using the given registry.
+func NewGatherer(registry Registry) *Gatherer {
+	return &Gatherer{
+		registry: registry,
+	}
+}
+
+// Gather gathers metrics from the registry and converts them to
+// a slice of metric families.
+func (g *Gatherer) Gather() (mfs []*dto.MetricFamily, err error) {
+	// Gather and pre-sort the metrics to avoid random listings
+	var names []string
+	g.registry.Each(func(name string, _ any) {
+		names = append(names, name)
+	})
+	slices.Sort(names)
+
+	mfs = make([]*dto.MetricFamily, 0, len(names))
+	for _, name := range names {
+		mf, err := metricFamily(g.registry, name)
+		switch {
+		case errors.Is(err, errMetricSkip):
+			continue
+		case err != nil:
+			return nil, err
```

Comment:
From the spec on `prometheus.Gatherer#Gather`:

> // Even if an error occurs, Gather attempts to gather as many metrics as
    // possible. Hence, if a non-nil error is returned, the returned
    // MetricFamily slice could be nil (in case of a fatal error that
    // prevented any meaningful metric collection) or contain a number of
    // MetricFamily protobufs, some of which might be incomplete, and some
    // might be missing altogether.

Rather than erroring here, it seems like we should be accumulating an `[]error` and then returning `errors.Join` of the resulting (potentially empty) slice.